gamedata/event_marathon.go

```diff
@@ -61,23 +61,29 @@ type EventMarathon struct {
 }
 
 func (em *EventMarathon) GetNextReward(eventPoint int32) (generic.Nullable[int32], generic.Nullable[client.Content]) {
-	// TODO(optimization): this can be a binary search
-	for _, reward := range em.TopStatus.EventMarathonPointRewardMasterRows.Slice {
-		if reward.RequiredPoint > eventPoint {
-			content := em.Gamedata.EventMarathonReward[reward.RewardGroupId][0]
-			return generic.NewNullable(reward.RequiredPoint), generic.NewNullableFromPointer(content)
-		}
+	slice := em.TopStatus.EventMarathonPointRewardMasterRows.Slice
+	idx := sort.Search(len(slice), func(i int) bool {
+		return slice[i].RequiredPoint > eventPoint
+	})
+
+	if idx < len(slice) {
+		content := em.Gamedata.EventMarathonReward[slice[idx].RequiredPoint][0]
+		return generic.NewNullable(slice[idx].RequiredPoint), generic.NewNullableFromPointer(content)
 	}
+
 	return generic.Nullable[int32]{}, generic.Nullable[client.Content]{}
 }
 
 func (em *EventMarathon) GetRankingReward(rank int32) int32 {
-	// TODO(optimization): this can be a binary search
-	for _, reward := range em.TopStatus.EventMarathonRankingRewardMasterRows.Slice {
-		if (!reward.LowerRank.HasValue) || (reward.LowerRank.Value >= rank) {
-			return reward.RewardGroupId
-		}
+	slice := em.TopStatus.EventMarathonRankingRewardMasterRows.Slice
```

Comment:
A funny thing is this would probably ended up being slower in practice, because most people will fall into the first ranking bracket anyway.

Reply:
That makes sense, thanks for clarification. I'm still reading up more about the code so feel free to correct anything that's off, I'll drop that change.

I'm assuming the other change for `em.TopStatus.EventMarathonPointRewardMasterRows.Slice` should be fine? These are all based on the assumption that the slices are sorted so that's what I'm still trying to test out

Reply:
yep, it's sorted based on increasing RequiredPoint.
generally, for the gamedata.Gamedata fields, you can find the loader in the file with similar name to the type (which is often also just the field name).